### ISSUE TITLE
🔢 Implement Matrix Multiplication (matmul)

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#14]
# 🔢 Implement Matrix Multiplication (matmul)

## 📋 Description

Implement efficient matrix multiplication functionality for tensors, supporting both 2D matrix multiplication and batch matrix multiplication. This is a fundamental operation needed for neural network layers and many tensor computations.

## 🎯 Acceptance Criteria

- [ ] Implement `matmul()` static method in Tensor class
- [ ] Support 2D matrix multiplication (M×N @ N×P = M×P)
- [ ] Support batch matrix multiplication for 3D+ tensors
- [ ] Add `@` operator overload for matrix multiplication
- [ ] Implement gradient computation for matmul operations
- [ ] Add comprehensive error checking for dimension compatibility
- [ ] Optimize for common cases (vector-matrix, matrix-vector)
- [ ] Support different data types with proper casting
- [ ] Add performance benchmarks against NumPy

### Matrix Multiplication Rules

**2D Case:**
```python
A: (m, k) @ B: (k, n) -> C: (m, n)
C[i, j] = sum(A[i, r] * B[r, j] for r in range(k))
```

**Batch Case:**
```python
A: (..., m, k) @ B: (..., k, n) -> C: (..., m, n)
# Broadcasting rules apply to batch dimensions
```

### Implementation Strategy

**Phase 1: Basic 2D Matrix Multiplication**
- [ ] Implement naive O(n³) algorithm
- [ ] Add dimension validation
- [ ] Handle edge cases (empty matrices, 1x1 matrices)

**Phase 2: Batch Operations**
- [ ] Support 3D+ tensor multiplication
- [ ] Implement batch dimension broadcasting
- [ ] Optimize batch operations

**Phase 3: Optimization**
- [ ] Block-based multiplication for cache efficiency
- [ ] Memory layout optimization
- [ ] Consider future BLAS integration

### Autograd Support

**MatMul Function Class:**
```python
class MatMul(Function):
    @staticmethod
    def forward(ctx, a, b):
        # Save inputs for backward pass
        ctx.save_for_backward(a, b)
        return matrix_multiply(a, b)
    
    @staticmethod
    def backward(ctx, grad_output):
        a, b = ctx.saved_tensors
        # da = grad_output @ b.T
        # db = a.T @ grad_output
        return grad_a, grad_b
```

## 🧪 Testing Requirements

- [ ] Test basic 2D matrix multiplication
- [ ] Test batch matrix multiplication
- [ ] Test edge cases (1x1, empty matrices)
- [ ] Test with different data types
- [ ] Verify gradients with numerical differentiation
- [ ] Performance benchmarks vs NumPy
- [ ] Memory usage profiling
- [ ] Test with very large matrices

## 📁 Files to Modify

- `grad/tensor.py` - Add matmul method and @ operator
- `grad/autograd/ops.py` - Add MatMul function class
- `grad/utils/linalg.py` - New file for linear algebra utilities
- `tests/test_matmul.py` - Comprehensive test suite
- `tests/test_performance.py` - Performance benchmarks

## 🔗 Dependencies

- Requires: Broadcasting support (#02)
- Requires: Autograd backward pass (#01)
- Requires: Gradient accumulation (#03)
- Blocks: Linear layers (#08)
- Blocks: Neural network implementations

### Basic Usage
```python
# 2D matrix multiplication
A = Tensor([[1, 2], [3, 4]], requires_grad=True)  # (2, 2)
B = Tensor([[5, 6], [7, 8]], requires_grad=True)  # (2, 2)

C = Tensor.matmul(A, B)  # or A @ B
# Result: [[19, 22], [43, 50]]

# Backward pass
loss = C.sum()
loss.backward()
print(A.grad)  # Gradient w.r.t A
print(B.grad)  # Gradient w.r.t B
```

### Batch Operations
```python
# Batch matrix multiplication
A = Tensor.randn(10, 64, 32)  # Batch of 10 matrices
B = Tensor.randn(10, 32, 128) # Batch of 10 matrices
C = A @ B                     # Result: (10, 64, 128)
```

### Vector-Matrix Operations
```python
# Vector-matrix multiplication
v = Tensor([1, 2, 3])         # (3,)
M = Tensor([[1, 2], [3, 4], [5, 6]])  # (3, 2)
result = v @ M                # (2,) - vector result
```

## 🚨 Edge Cases to Handle

- Matrix dimension mismatches
- Empty matrices (0-dimensional)
- Very large matrices (memory constraints)
- Mixed data types
- Non-contiguous tensors
- Batch dimension broadcasting
- Vector-matrix and matrix-vector cases

## 🔧 Performance Considerations

- Cache-friendly memory access patterns
- Minimize temporary memory allocation
- Efficient handling of non-contiguous tensors
- Consider future GPU implementation
- Memory pooling for intermediate results

## 🏷️ Labels

- `enhancement`
- `linear-algebra`
- `core-functionality`
- `performance`
- `high-priority`

## 📈 Priority

**High** - Essential for neural networks and most ML algorithms

## 🔗 Related Issues

- Depends on: Broadcasting (#02)
- Depends on: Autograd (#01, #03)
- Blocks: Linear layers (#08)
- Blocks: Neural network training (#11)
- Related to: Performance optimization
